src/zulip/api.rs

```diff
@@ -0,0 +1,201 @@
+            return Ok(());
+        }
+
+        let member_ids = format!(
+            "[{}]",
+            member_ids
+                .iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>()
+                .join(",")
+        );
+        let mut form = HashMap::new();
+        form.insert("name", user_group_name);
+        form.insert("description", description);
+        form.insert("members", &member_ids);
+
+        let mut r = self.req(reqwest::Method::POST, "/user_groups/create", Some(form))?;
+        if r.status() == 400 {
+            let body = r.json::<serde_json::Value>()?;
+            let err = || {
+                failure::format_err!(
+                    "got 400 when creating user group {}: {}",
+                    user_group_name,
+                    body
+                )
+            };
+            let error = body.get("msg").ok_or_else(err)?.as_str().ok_or_else(err)?;
+            if error.contains("already exists") {
+                return Ok(());
+            } else {
+                return Err(err());
+            }
+        }
+
+        r.error_for_status()?;
+
+        Ok(())
+    }
+
+    /// Get all user groups of the Rust Zulip instance
+    pub(crate) fn get_user_groups(&self) -> Result<Vec<ZulipUserGroup>, Error> {
+        let response = self
+            .req(reqwest::Method::GET, "/user_groups", None)?
+            .error_for_status()?
+            .json::<ZulipUserGroups>()?
+            .user_groups;
+
+        Ok(response)
+    }
+
+    /// Get all users of the Rust Zulip instance
+    pub(crate) fn get_users(&self) -> Result<Vec<ZulipUser>, Error> {
+        let response = self
+            .req(reqwest::Method::GET, "/users", None)?
+            .error_for_status()?
+            .json::<ZulipUsers>()?
+            .members;
+
+        Ok(response)
+    }
+
+    pub(crate) fn update_user_group_members(
+        &self,
+        user_group_id: usize,
+        add_ids: &[usize],
+        remove_ids: &[usize],
+    ) -> Result<(), Error> {
+        if add_ids.is_empty() && remove_ids.is_empty() {
+            log::info!(
+                "user group {} does not need to have its group members updated",
+                user_group_id
+            );
+            return Ok(());
+        }
+
+        log::info!(
+            "updating user group {} by adding {:?} and removing {:?}",
+            user_group_id,
+            add_ids,
+            remove_ids
+        );
+
+        if self.dry_run {
+            return Ok(());
+        }
+
+        let add_ids = format!(
+            "[{}]",
+            add_ids
+                .iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>()
+                .join(",")
+        );
+        let remove_ids = format!(
+            "[{}]",
+            remove_ids
+                .iter()
+                .map(|id| id.to_string())
+                .collect::<Vec<_>>()
+                .join(",")
+        );
```

Comment:
Nit: this is fairly noisy and has been repeated a lot, factoring out would make the code easier to read :)